### PR TITLE
Add support for LP5562 LED driver

### PR DESCRIPTION
**Pull request by peterbarker**
This driver costs 472 bytes when compiled into KakuteH7 with this extra hwdef:
```
pbarker@fx:~$ cat /tmp/extra.hwdef 
define AP_NOTIFY_LP5562_BUS 1
define AP_NOTIFY_LP5562_ADDR 0x60
pbarker@fx:~$ 
```

This works with the supplied simulator.

Yet to be tested on hardware.

### CODE REVIEW

**Comment by tridge** [comment]
The LED is working on my Pixhawk6X with: ```
define AP_NOTIFY_LP5562_BUS 3
define AP_NOTIFY_LP5562_ADDR 0x30
```

File: Tools/ardupilotwaf/boards.py [hunk 1]
```diff
@@ -642,6 +642,9 @@ def configure_env(self, cfg, env):
         cfg.define('AP_OPENDRONEID_ENABLED', 1)
         cfg.define('AP_SIGNED_FIRMWARE', 0)
 
+        cfg.define('AP_NOTIFY_LP5562_BUS', 2)
+        cfg.define('AP_NOTIFY_LP5562_ADDR', 0x66)
```
Comment (tridge) [inline Tools/ardupilotwaf/boards.py:646]: change to 0x30

File: libraries/AP_Notify/LP5562.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+enum class Register {
+    ENABLE  = 0x00,
+    B_PWM   = 0x02,
+    G_PWM   = 0x03,
+    R_PWM   = 0x04,
+    CONFIG  = 0x08,
+    LED_MAP = 0x70,
+};
+
+LP5562::LP5562(uint8_t bus, uint8_t addr)
```
Comment (tridge) [inline libraries/AP_Notify/LP5562.cpp:45]: maybe we should make addr 0 probe for all 4 possible addresses, 0x30, 0x31, 0x32 and 0x33

File: libraries/SITL/SIM_I2C.cpp [hunk 3]
```diff
@@ -85,6 +87,7 @@ struct i2c_device_at_address {
     { 2, 0x0B, maxell },        // Maxell: BATTx_MONITOR 16, BATTx_I2C_ADDR 13
     { 3, 0x0B, smbus_generic},  // BATTx_MONITOR 7, BATTx_I2C_ADDR 13
     { 2, 0x28, airspeed_dlvr }, // ARSPD_TYPE = 7 5inch H2O sensor
+    { 2, 0x66, lp5562 },        // LP5562 RGB LED driver
```
Comment (tridge) [inline libraries/SITL/SIM_I2C.cpp:90]: should be 0x30

File: libraries/AP_Notify/LP5562.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    _dev->set_retries(10);
+
+    //  values here are taken literally from 7.3.2 in the datasheet.
+    //  See the simulator for register breakdown.
+
+    // chip enable:
+    _dev->write_register((uint8_t)Register::ENABLE, 0b1000000);
+
+    // start-up-delay:
+    hal.scheduler->delay_microseconds(500);
```
Comment (tridge) [inline libraries/AP_Notify/LP5562.cpp:69]: is that delay really needed?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Notify/LP5562.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_Notify/LP5562.cpp:69]: I tried on mine and it works without the delay

Comment (peterbarker) [reply libraries/AP_Notify/LP5562.cpp:69]: According to the datasheet, yes!

I haven't tested if it is needed in reality.

File: libraries/AP_Notify/LP5562.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_Notify/LP5562.cpp:45]: We can add that later when someone asks for it :-)

Until then we'd just be wasting bytes.

File: Tools/ardupilotwaf/boards.py (hunk 1 above)
Comment (peterbarker) [reply Tools/ardupilotwaf/boards.py]: Done

File: libraries/SITL/SIM_I2C.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/SITL/SIM_I2C.cpp]: Fixed

File: libraries/AP_Notify/LP5562.cpp (hunk 4 above)
Comment (peterbarker) [reply libraries/AP_Notify/LP5562.cpp:69]: > I tried on mine and it works without the delay

Do we take a punt and remove the delay?

A third option would be to add a `bool _initialised` to the class and do all of the setup work in the periodic callback instead.

Comment (tridge) [reply libraries/AP_Notify/LP5562.cpp:69]: if the datasheet says its needed we should leave it in

File: libraries/AP_Notify/LP5562.cpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    // use internal clock:
+    _dev->write_register((uint8_t)Register::CONFIG, 0b00000001);
+
+    // set direct PWM control:
+    _dev->write_register((uint8_t)Register::LED_MAP, 0b00000000);
+
+    _dev->set_retries(1);
+
+    _dev->register_periodic_callback(20000, FUNCTOR_BIND_MEMBER(&LP5562::_timer, void));
```
Comment (tridge) [inline libraries/AP_Notify/LP5562.cpp:79]: we need to check the write_register calls, and not install the callback unless they succeed
even better would be a detection that this really is a LP5562, perhaps do a RESET then read LED_MAP and check for 0x71?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/Pixhawk6X/hwdef.dat [hunk 6]
```diff
@@ -366,3 +366,5 @@ ENABLE_DFU_BOOT 1
 
 # build ABIN for flash-from-bootloader support:
 env BUILD_ABIN True
+
+define AP_NOTIFY_LP5562_ENABLED 1
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/Pixhawk6X/hwdef.dat:370]: needed for at least the Pixhawk6C too, and likely other boards

**Review by tridge — APPROVED** [review]
I've tested and the updated driver works well on both I2C and GPS ports on Pixhawk6X

**Comment by tridge** [comment]
only question is if we need to find some way to make this work without users having to update params

**Comment by tridge** [comment]
@peterbarker also, is dimming supposed to work? NTF_LED_BRIGHT=0 does turn it off, but I can't see any change between levels 1 to 3

File: libraries/AP_Notify/AP_Notify.h [hunk 7]
```diff
@@ -83,6 +83,10 @@ class AP_Notify
         Notify_LED_Scripting                = (1 << 10),// Colour accessor for scripting
         Notify_LED_DShot                    = (1 << 11),// Use dshot commands to set ESC LEDs
         Notify_LED_ProfiLED_SPI             = (1 << 12), // ProfiLED
+#if AP_NOTIFY_LP5562_ENABLED
```
Comment (tridge) [inline libraries/AP_Notify/AP_Notify.h:86]: needs to be on all boards

**Review by tridge — CHANGES_REQUESTED** [review]
we need probe checks for each of our I2C LEDs, to make sure the device really exists

**Review by tridge — APPROVED** [review]

**Pull request merged**